includes/modules/core/command/PrivmsgCommand.hpp

```diff
@@ -15,22 +15,31 @@
 class PrivmsgCommand : public CommandExecutor {
 
 	void execute(const Command& cmd, CommandSender& sender) {
-		User& user = static_cast<User&>(sender);
-
-        vector<CommandSender *> target = cmd.getArg<vector<CommandSender *> >("msgtarget");
-        string& message = cmd.getArg<string &>("message");
+		User &user = static_cast<User &>(sender);
+		string from = user.getName() + "!" + user.getUserName() + "@" + Irc::getInstance().getServer().getHost();;
+		vector<CommandSender *>& target = cmd.getArg<vector<CommandSender *> >("msgtarget");
+		string message = cmd.getArg<string>("message") + "\n";
 
 		for (vector<CommandSender *>::iterator it = target.begin(); it != target.end(); it++) {
-            try {
-                Channel *channel = dynamic_cast<Channel *>(*it);
-                if (!channel->isOnChan(static_cast<User *>(&sender))) {
-                    sender.send(ResponseTypes::ERR_NOTONCHANNEL((*it)->getName().c_str(), sender.getName().c_str()));
-                    continue ;
-                }
-            } catch(std::bad_cast& e){}
-			(*it)->send(message.c_str());
+			if (Channel::isValidIdentifier((*it)->getName())) {
+				if (Irc::getInstance().getChannels().find((*it)->getName()) != Irc::getInstance().getChannels().end()) {
+					if (!Irc::getInstance().getChannels().find((*it)->getName())->second->isOnChan(static_cast<User *>(&sender)) && Irc::getInstance().getChannels().find((*it)->getName())->second->getFlag().outside_message) {
```

Comment:
A user variable is already declared.
```suggestion
					if (!Irc::getInstance().getChannels().find((*it)->getName())->second->isOnChan(&user) && Irc::getInstance().getChannels().find((*it)->getName())->second->getFlag().outside_message) {
```